igv_remote/_core.py

```diff
@@ -70,42 +85,35 @@ def _load(self, *urls):
 
 
     def _adjust_viewopts(self):
+        
         # specify view options
-        if self._squish:
+        if self._view_type == "squished":
             self._send( "squish ")
-        if self._collapse:
+        elif self._view_type == "collapsed":
             self._send( "collapse ")
+        elif self._view_type == "expanded":
+            self._send("expand ")
+        else:
+            print("view_type other than squished/collapsed/expanded cannot be understood, will use expand")
+            self._send("expand ")
+
         if self._viewaspairs:
             self._send( "viewaspairs ")
         self._send( "sort {}".format(self._sort))
     
-    
-    
     def _goto(self, 
-             chromosome=None, pos=None, end_pos=None):
-        """
-        Note that the position params could be either list or single element.
-        examples: chromosome=2, start_pos=[34,3890,34859], end_pos = [3544, 6909, 34980]
-        """
-        
-        if end_pos is None:
-            
-            start_pos = pos-1
-            end_pos = start_pos+2
-            #position = "chr{chr}:{pos}".format(chr=chromosome, pos=pos)
-        elif start_pos and end_pos:
-            start_pos ='{:,}'.format(start_pos)
-            end_pos = '{:,}'.format(end_pos)
-            
-        else:
-            raise Exception("No view location specified")
-        
-        position= 'chr{}:{}-{}'.format(chromosome,start_pos, end_pos)
+             chromosome=None, pos1=None, pos2=None, expand=20):
+
+        position = _parse_loc(chromosome, pos1, pos2, expand)
         print("position to view:", position)
 
         self._send( "goto %s" % position)
-        self._adjust_viewopts()
-        
+    
+    def _goto_multiple(self, expand=20, **loci):
```

Comment:
Why use kwargs here? There are two problems with this:

1. Since all arguments are mandatory, it's good to be explicit defining them in the function signature
2. There are no checks on whether the keys are present. If any single argument in `chr1`, `pos1`, `chr2`, `pos2` is not given, the tool will raise a cryptic `KeyError`.

Most of the time, kwargs is used for passing through arbitrary arguments to another function, when they are irrelevant to the outer function, e.g.:

```python
def foo(a, b, **kwargs)
  do_something_with_named_args(a, b)
  bar(**kwargs)

def bar(x, y):
   do_something(x, y)

foo(a = 1, b = 2, x = 2, y = 4)
```

Since `foo` does not care about x and y, it's convenient to wrap them in a single kwargs to pass to bar.

When using it in your case (i.e. as the primary way to supply arguments), it's generally used if a function can admit either a huge list of arguments or a dynamic list of arguments, and it's easy to programatically access them/check if they're valid, e.g.

```python
def foo(**kwargs):
  for k, v in kwargs.items():
    validate_param(k, v)
```

So for this function, I'd update it to

```python
def _goto_multiple(self, chr1, pos1, chr2, pos2, expand=20, ):
  self._send("goto {} {}".format(*[_parse_loc(c, p, None, expand), for c, p in [[chr1, pos1], [chr2, pos2]]]))
```

This both eliminates the kwargs and also makes the format string a bit more compact; if you ever decide to redefine `_parse_loc` you only need to edit its call in a single place, rather than two.

Reply:
I am not sure if anyone wants more than two loci (e.g. if someone wants `chr3, pos3, ...`)?

Reply:
If you want to support arbitrarily many loci, then kwargs would be the way to go. But you'll need to add a step that verifies the arguments, e.g.

```python
chrpos = { "chr" : {}, "pos" : {} }

try:
  for k, v in kwargs.items():
    arg = re.match(r'^(chr|pos)(\d+)$', k)
    if arg is not None:
      chrpos[arg[1]][arg[2]] = v
    else:
      raise Exception
  if chrpos["chr"].keys() != chrpos["pos"].keys():
    raise Exception
except:
  raise ValueError("When specifying multiple loci, arguments must be of the format chr1 = <chr1>, pos1 = <pos1>, ..., chrN = <chrN>, posN = <posN>")
```

I'll leave it up to you to figure out the best way to construct the `goto` command.

Reply:
Very nice! Thanks for this suggestion with regex!